forge/src/swarm/SwarmV2.sol

```diff
@@ -0,0 +1,422 @@
+    }
+
+    modifier onlyAggregator(address aggregator) {
+        if (!IAccessControl(getAccessControl()).isAggregator(aggregator)) {
+            revert NotAggregator();
+        }
+        _;
+    }
+
+    modifier onlyTrainer(address trainer) {
+        if (!IAccessControl(getAccessControl()).isTrainer(trainer)) {
+            revert NotTrainer();
+        }
+        _;
+    }
+
+    modifier onlyEvaluator(address evaluator) {
+        if (!IAccessControl(getAccessControl()).isEvaluator(evaluator)) {
+            revert NotEvaluator();
+        }
+        _;
+    }
+
+    function initialize(
+        SwarmV2InitializeParams memory params
+    ) external virtual initializer {
+        __EIP712_init(params.name, _VERSION);
+        __RoundTraining_init();
+        __BaseTrainingPhases_init(params.trainingPhaseConfiguration, params.evaluationPhaseConfiguration);
+        __CertificateRegistry_init();
+        __RoundTrainerRegistry_init();
+        __RoundEvaluatorRegistry_init();
+        __TaskAssignment_init();
+        __SwarmCore_init(params.initialTrainerSelector, params.initialEvaluatorSelector, params.initialContributionCalculator, params.initialAccessControl, params.initialCompensation);
+    }
+
+    function initialize() external virtual override(RoundTrainerRegistryV2, RoundEvaluatorRegistry, TaskAssignment) {
+        revert WrongInitialization();
+    }
+
+    function canTrain(
+        address trainer,
+        uint256 roundId
+    ) public view returns (bool) {
+        ISelector selector = ISelector(getTrainerSelector());
+        IAccessControl accessControl = IAccessControl(getAccessControl());
+        return accessControl.isTrainer(trainer) && selector.isSelected(trainer, roundId);
+    }
+
+    function canEvaluate(
+        address evaluator,
+        uint256 roundId
+    ) public view returns (bool) {
+        ISelector selector = ISelector(getEvaluatorSelector());
+        IAccessControl accessControl = IAccessControl(getAccessControl());
+        return accessControl.isEvaluator(evaluator) && selector.isSelected(evaluator, roundId);
+    }
+
+    function updateTrainerSelector(address newTrainerSelector) external onlyAggregator(msg.sender) {
+        _updateTrainerSelector(newTrainerSelector);
+    }
+
+    function updateEvaluatorSelector(address newEvaluatorSelector) external onlyAggregator(msg.sender) {
+        _updateEvaluatorSelector(newEvaluatorSelector);
+    }
+
+    bool private _trainerPrivacyEnabled;
+
+    function distribute(
+        uint256 roundId,
+        address[] calldata trainers,
+        uint64[] calldata contributions
+    ) external onlyAggregator(msg.sender) {
+        _distribute(roundId, trainers, contributions);
+    }
+
+    function registerRoundContributionPrivacy(uint256 roundId, bytes32 commitment, bytes32 modelHash, uint64 revealDeadline) external onlyAggregator(msg.sender) {
+        if (!_trainerPrivacyEnabled) {
+            revert PrivacyModeDisabled();
```

Comment:
This feature around _trainerPrivacyEnabled doesn't seem required and adds a bit of extra steps and code. If the objective is to wait for the `configureTrainerPrivacy` and  `setTrainerPrivacyMode` to be called before starting training, I believe you don't need this.

The flow should simply be:
1. (already implemented) The swarm starts in IDLE_PHASE. Training is not started as it reverts on the next if. 
2. call `configureTrainerPrivacy`
3. (already implemented) call `startTrainingRound`